### PR TITLE
Cat Burglar support, slightly more XO Skeleton hugging

### PR DESCRIPTION
**Pull request by jaspercb**
Totally untested, and I find it very believable that there's a bug in this that causes us to burn two catburgles on an item mistakenly. Please read carefully and do not merge.

### CODE REVIEW

File: RELEASE/scripts/sl_ascend/sl_mr2018.ash [hunk 1]
```diff
@@ -440,7 +440,7 @@ boolean songboomSetting(int option)
 
 int catBurglarHeistsLeft()
 {
-	if (!have_familiar($familiar[Cat Burglar]) !sl_is_valid($familiar[Cat Burglar]) || !is_unrestricted($item[kitten burglar]))
+	if (!have_familiar($familiar[Cat Burglar]) || !sl_is_valid($familiar[Cat Burglar]) || !is_unrestricted($item[kitten burglar]))
```
Comment (soolar) [inline RELEASE/scripts/sl_ascend/sl_mr2018.ash]: `sl_is_valid` already checks `is_unrestricted`, so no need to do that here as well.

File: RELEASE/scripts/sl_ascend/sl_combat.ash [hunk 2]
```diff
@@ -986,6 +986,14 @@ string sl_combatHandler(int round, string opp, string text)
 			// snatch a hedge trimmer if we're lucky
 			if($monsters[bearpig topiary animal, elephant (meatcar?) topiary animal, spider (duck?) topiary animal] contains enemy)
 				dohug = true;
+
+			// snatch a killing jar if we're lucky
+			if(enemy == $monster[banshee librarian] && (0 ==item_amount($item[Killing jar])))
```
Comment (soolar) [inline RELEASE/scripts/sl_ascend/sl_combat.ash]: super nitpick: space between `==` and `item_amount`

**Comment by jaspercb** [comment]
Nits fixed.

We actually already run the Cat Burglar over the XO skeleton until we get the first two heists each day!

My test plan for this is roughly "add wishing for mountain man in hardcore, do a full 3-day HC ascension, confirm that we wished for mountain man and that all burglar uses were reasonable". So turnaround time will be high.

```
if((get_property("_catBurglarCharge").to_int() < 30) && (estimatedTurnsLeft() > (30 - get_property("_catBurglarCharge").to_int())))
		{
			fams = ListRemove(fams, $familiar[Cat Burglar]);
			fams = ListInsertAt(fams, $familiar[Cat Burglar], fams.ListFind($familiar[XO Skeleton]));
		}
```

**Comment by jaspercb** [comment]
Mountain man:
```
> Turn(180): Starting with 3 left at Level: 10
> Encounter: -26.0   Exp Bonus: 18.945833333333333
> Meat Drop: 300.38699100999077	 Item Drop: 205.19349550499538
> HP: 200/200, MP: 751/857
> Tummy: 15/15 Liver: 8/15 Spleen: 4/15
> ML: 50 control: 10
> Soulsauce: 41
> Everything Looks Yellow: 88
> Delay between adventures... beep boop... 
> I think I'm paranoid and complicated
> I think I'm paranoid, manipulated
Preference sl_paranoia_counter changed from 2 to 0
Preference sl_familiarChoice changed from Cat Burglar to XO Skeleton

equip off-hand Kramco Sausage-o-Matic™
> want to heist a bird rib from a whitesnake
> want to heist a lion oil from a white lion
> want to heist a amulet of extreme plot significance from a Quiet Healer
> want to heist a chrome ore from a mountain man
> Not eating or drinking anything, since we don't know what's good...
> Trying to wish for a mountain man, which the cat will then burgle, hopefully.
Preference sl_familiarChoice changed from XO Skeleton to Cat Burglar
Preference nextAdventure changed from The Penultimate Fantasy Airship to Noob Cave
> Starting preadventure script...
> want to heist a bird rib from a whitesnake
> want to heist a lion oil from a white lion
> want to heist a amulet of extreme plot significance from a Quiet Healer
> want to heist a chrome ore from a mountain man
Preference sl_priorLocation changed from The Penultimate Fantasy Airship to Noob Cave
> Pre Adventure at Noob Cave done, beep.
> About to start a combat indirectly at Noob Cave... (3) accesses required.

use 1 pocket wish
Took choice 1267/1: (secret choice)
choice.php?pwd8c27eff3a79a46c141f7295e624ace66=&whichchoice=1267&option=1&wish=to+fight+a+mountain+man
Preference _genieWishesUsed changed from 1 to 2
Preference _genieFightsUsed changed from 0 to 1
Preference nextAdventure changed from Noob Cave to None

[181] genie summoned monster
Preference lastEncounter changed from Irritating Series of Random Encounters to mountain man
Encounter: mountain man
Preference _lastCombatStarted changed from 20190527233841 to 20190527234153
Round 0: Jeparo wins initiative!
> slAdvBypass has encountered a combat! (param: 'sl_combatHandler')
Preference nextAdventure changed from None to Noob Cave
Preference lastAdventure changed from The Penultimate Fantasy Airship to Noob Cave

[181] Noob Cave
Encounter: mountain man
Round 0: Jeparo loses initiative!
> sl_combatHandler: 0
Preference sl_combatHandler changed from  to (yellowray)
Preference sl_yellowRays changed from (1:Burly Sidekick:Disintegrate:168) to (1:Burly Sidekick:Disintegrate:168), (1:mountain man:Asdon Martin: Missile Launcher:180)
Preference _missileLauncherUsed changed from false to true
Round 1: Jeparo casts ASDON MARTIN: MISSILE LAUNCHER!
Round 2: Jeparo wins the fight!
You gain 14 Meat.
After Battle: Treacherous Donskoy does a dance so sneaky, it looks like he's just standing still.
You acquire an item: asbestos ore
You acquire an item: asbestos ore
You acquire an item: linoleum ore
You acquire an item: linoleum ore
You acquire an item: chrome ore
You acquire an item: chrome ore
After Battle: You gain 7 Strongness
After Battle: You gain 21 Mysteriousness
After Battle: You gain 14 Sarcasm
You gain a Moxie point!
After Battle: Treacherous Donskoy takes note of any security cameras in the area.
You gain 3 Soulsauce
Preference testudinalTeachings changed from 213:4|171:0|214:5|267:2|72:2|-1:5|209:3|168:2|266:3 to 213:4|171:0|214:5|267:3|72:2|-1:5|209:3|168:2|266:3
Preference _catBurglarCharge changed from 44 to 45
This combat did not cost a turn
Preference lastAdventure changed from Noob Cave to The Penultimate Fantasy Airship

cast 3 Soul Food
You gain 45 Mana Points

cast 1 Fat Leon's Phat Loot Lyric
You acquire an effect: Fat Leon's Phat Loot Lyric (10)

cast 1 Singer's Faithful Ocelot
You acquire an effect: Singer's Faithful Ocelot (10)

cast 1 Spiky Shell
You acquire an effect: Spiky Shell (5)
> Post Adventure done, beep.
> Turn(180): Starting with 3 left at Level: 10
> Encounter: -20.0   Exp Bonus: 14.133333333333335
> Meat Drop: 300.38699100999077	 Item Drop: 215.19349550499538
> HP: 200/200, MP: 762/1020
> Tummy: 15/15 Liver: 8/15 Spleen: 4/15
> ML: 50 control: 10
> Soulsauce: 29
> Everything Looks Yellow: 88
> Delay between adventures... beep boop... 
Preference sl_paranoia_counter changed from 0 to 1
Preference sl_familiarChoice changed from Cat Burglar to XO Skeleton

equip off-hand latte lovers member's mug

buy 8 soda water for 66 each from The General Store
You spent 528 Meat
You acquire soda water (8)

[181] Cook 8 wad of dough + 8 soda water
You acquire loaf of soda bread (8)
Crafting used 8 each of wad of dough and soda water

Converting 8 loaf of soda breads into Fuel
> Compressed 8 loaf of soda bread into sheep, I mean fuel: 41 --> 89
> want to heist a bird rib from a whitesnake
> want to heist a lion oil from a white lion
> want to heist a amulet of extreme plot significance from a Quiet Healer
> want to heist a chrome ore from a mountain man
> Trying to heist a chrome ore

main.php?heist=1
Took choice 1320/1: asbestos ore
choice.php?whichchoice=1320&option=1&st:1153:365=chrome+ore&pwd
You acquire an item: chrome ore
Preference _catBurglarHeistsComplete changed from 0 to 1
> Turn(180): Starting with 3 left at Level: 10
> Encounter: -26.0   Exp Bonus: 15.133333333333335
> Meat Drop: 315.82677881933927	 Item Drop: 212.91338940966963
> HP: 200/200, MP: 762/857
> Tummy: 15/15 Liver: 8/15 Spleen: 4/15
> ML: 50 control: 10
> Soulsauce: 29
> Everything Looks Yellow: 88
> Delay between adventures... beep boop... 
Preference sl_paranoia_counter changed from 1 to 2
> want to heist a bird rib from a whitesnake
> want to heist a lion oil from a white lion
> want to heist a amulet of extreme plot significance from a Quiet Healer
> Not eating or drinking anything, since we don't know what's good...
> Giving Trapper goat cheese and chrome ore
Preference sl_trapper changed from start to yeti

Visiting the Trapper
Preference questL08Trapper changed from step1 to step2
```

Amulet (thankfully not needed):
```
> Starting preadventure script...

familiar XO Skeleton (39 lbs)

Unequip Cat Burglar

equip familiar astral pet sweater
> want to heist a bird rib from a whitesnake
> want to heist a lion oil from a white lion
> want to heist a amulet of extreme plot significance from a Quiet Healer
> Using cat burglar because we want to burgle a amulet of extreme plot significance from Quiet Healer

familiar Cat Burglar (34 lbs)

Unequip XO Skeleton

equip familiar astral pet sweater
> We can't cap this drop bear!
> Pre Adventure at The Penultimate Fantasy Airship done, beep.

[197] The Penultimate Fantasy Airship
Preference lastEncounter changed from sausage goblin to Quiet Healer
Encounter: Quiet Healer
Preference _lastCombatStarted changed from 20190527235122 to 20190527235137
Round 0: Jeparo wins initiative!
> sl_combatHandler: 0
Preference sl_combatHP changed from 248 to 254
Preference sl_sniffs changed from (1:writing desk:Transcendent Olfaction:130), (1:writing desk:Gallapagosian Mating Call:130), (1:writing desk:Offer Latte to Opponent:130), (1:dairy goat:Gallapagosian Mating Call:159), (1:dairy goat:Offer Latte to Opponent:159), (1:Quiet Healer:Gallapagosian Mating Call:165) to (1:writing desk:Transcendent Olfaction:130), (1:writing desk:Gallapagosian Mating Call:130), (1:writing desk:Offer Latte to Opponent:130), (1:dairy goat:Gallapagosian Mating Call:159), (1:dairy goat:Offer Latte to Opponent:159), (1:Quiet Healer:Gallapagosian Mating Call:165), (1:Quiet Healer:Transcendent Olfaction:196)
Preference sl_combatHandler changed from  to (sk19)
Round 1: Jeparo casts TRANSCENDENT OLFACTION!
You acquire an effect: On the Trail (40)
Round 2: Quiet Healer takes 4 damage.
Round 2: Quiet Healer takes 1 damage.
Round 2: You lose 9 hit points
Preference olfactedMonster changed from writing desk to Quiet Healer
Preference sl_combatHP changed from 254 to 245
Preference sl_diag_round changed from 0 to 1
Preference sl_combatHandler changed from (sk19) to (sk19)(banishercheck)
Preference sl_combatHandler changed from (sk19)(banishercheck) to (sk19)(banishercheck)(sk4034)
Round 2: Jeparo casts CURSE OF WEAKSAUCE!
Round 3: Quiet Healer drops 5 attack power.
Round 3: Quiet Healer drops 6 defense.
Preference sl_diag_round changed from 1 to 2
Preference sl_combatHandler changed from (sk19)(banishercheck)(sk4034) to (sk19)(banishercheck)(sk4034)(sk7289)
Round 3: Jeparo casts MICROMETEORITE!
Round 4: Quiet Healer drops 12 attack power.
Round 4: Quiet Healer drops 11 defense.
Round 4: Quiet Healer drops 7 attack power.
Round 4: Quiet Healer drops 6 defense.
Preference _micrometeoriteUses changed from 179 to 180
Preference sl_diag_round changed from 2 to 3
Preference sl_combatHandler changed from (sk19)(banishercheck)(sk4034)(sk7289) to (sk19)(banishercheck)(sk4034)(sk7289)(sk7297)
Round 4: Jeparo casts SING ALONG!
Round 5: Quiet Healer drops 5 attack power.
Round 5: Quiet Healer drops 6 defense.
Preference sl_diag_round changed from 3 to 4
Preference sl_combatHandler changed from (sk19)(banishercheck)(sk4034)(sk7289)(sk7297) to (sk19)(banishercheck)(sk4034)(sk7289)(sk7297)(sk3007)
Round 5: Jeparo casts STUFFED MORTAR SHELL!
Round 6: Quiet Healer drops 7 attack power.
Round 6: Quiet Healer drops 5 defense.
Preference sl_diag_round changed from 4 to 5
Preference sl_combatHandler changed from (sk19)(banishercheck)(sk4034)(sk7289)(sk7297)(sk3007) to (sk19)(banishercheck)(sk4034)(sk7289)(sk7297)(sk3007)(sk7182)
Preference sl_combatHandler changed from (sk19)(banishercheck)(sk4034)(sk7289)(sk7297)(sk3007)(sk7182) to (sk19)(banishercheck)(sk4034)(sk7289)(sk7297)(sk3007)(delaymortarshell)
Round 6: Jeparo casts SALSABALL!
Round 7: Quiet Healer takes 8 damage.
Round 7: Quiet Healer takes 300 damage.
Round 7: Quiet Healer drops 7 attack power.
Round 7: Quiet Healer drops 7 defense.
Round 7: Jeparo wins the fight!
After Battle: You gain 50 Mana Points
You gain 564 Meat
You gain 10 Meat.
After Battle: Treacherous Donskoy shoots you a wink and does a little dance to disabuse onlookers of his nefarious intentions.
You acquire an item: amulet of extreme plot significance
You acquire an item: scroll of drastic healing
You acquire an item: soft green echo eyedrop antidote
After Battle: You gain 11 Fortitude
After Battle: You gain 17 Magicalness
After Battle: You gain 16 Cheek
You gain a Moxie point!
After Battle: Treacherous Donskoy takes note of any security cameras in the area.
You gain 4 Soulsauce
Preference testudinalTeachings changed from 213:2|171:0|214:5|267:5|72:2|-1:5|209:3|168:2|266:3|69:0 to 213:2|171:0|214:5|267:0|72:2|-1:5|209:3|168:2|266:3|69:0
Preference _catBurglarCharge changed from 47 to 48
Preference _boomBoxFights changed from 0 to 1
```

File: RELEASE/scripts/presool.ash [hunk 3]
```diff
@@ -59,6 +59,27 @@ void handlePreAdventure(location place)
 		}
 	}
 
+	if(sl_have_familiar($familiar[cat burglar]) && sl_is_valid($familiar[cat burglar]))
```
Comment (soolar) [inline RELEASE/scripts/presool.ash]: `sl_have_familiar` includes an `sl_is_valid` check already.

File: RELEASE/scripts/sl_ascend.ash [hunk 4]
```diff
@@ -7387,6 +7374,10 @@ boolean L12_gremlins()
 		}
 	}
 
+	if(0 < have_effect($effect[Curse of the Black Pearl Onion])) {
+		uneffect($effect[Curse of the Black Pearl Onion]);
```
Comment (soolar) [inline RELEASE/scripts/sl_ascend.ash]: What's the deal with this? Something tcrs related?

File: RELEASE/scripts/sl_ascend/sl_mr2017.ash [hunk 5]
```diff
@@ -1675,15 +1675,15 @@ boolean makeGenieCombat(monster mon, string option)
...
+	string wish = "to fight a " + mon;
 	string[int] pages;
 	int wish_provider = 9537; // pocket wish
 	if(get_property("_genieWishesUsed").to_int() < 3)
 	{
 		int wish_provider = 9529; // genie bottle
 	}
 	pages[0] = "inv_use.php?pwd=" + my_hash() + "&which=3&whichitem="+wish_provider;		//false
-	pages[1] = "choice.php?pwd=&whichchoice=1267&option=1&wish=" + wish;
+	pages[1] = "choice.php?pwd" + my_hash() + "=&whichchoice=1267&option=1&wish=" + wish;
```
Comment (soolar) [inline RELEASE/scripts/sl_ascend/sl_mr2017.ash]: `=` should be before `my_hash()`

File: RELEASE/scripts/sl_ascend.ash (hunk 4 above)
Comment (jaspercb) [reply RELEASE/scripts/sl_ascend.ash]: Whoops, forgot to remove from my personal stack of changes. Yeah, it's given by Rhinestones in Sauceror/Blender, makes combat vastly easier, but obviously ruins gremlins. No harm in leaving it in, I suppose.

File: RELEASE/scripts/presool.ash (hunk 3 above)
Comment (jaspercb) [reply RELEASE/scripts/presool.ash]: ugh this again sorry yes

File: RELEASE/scripts/sl_ascend/sl_mr2017.ash (hunk 5 above)
Comment (jaspercb) [reply RELEASE/scripts/sl_ascend/sl_mr2017.ash]: :|

**Pull request merged**